### ISSUE TITLE
Job to download AHL for Appeals

### ISSUE THREAD

**Issue by mdbenjam** [department-of-veterans-affairs/caseflow#8414]
# Problem
Before a Veteran can be scheduled for a hearing, we need to alert the hearing coordinator of any possible alternative hearing locations (AHL). Determining the AHLs is slow (about 1-2 seconds per Veteran). Unfortunately there are a lot of Veterans that need to be scheduled at once, making the requests to get the AHL for all these Veterans will take on the order of minutes. This means we cannot make all of the requests on page load.

# Solution
Background job. We can write a background job that gets the AHL for all of the Veterans' that need to be scheduled. We can save these AHL in a DB table so that they're easy to query.

## AHL Table
Create a new `hearing_location` table. Each `appeal` can have multiple records in the `hearing_location` table. Each record will have the following columns.

- `distance` 
- `facility_id`
- `name`
- `address`

## Job
We should run the job on a cron every hour for a limited set of Veterans. Running it frequently for small numbers of Veterans has some advantages
1) We are rate limited, so spreading the load out will allow us to avoid hitting these limits.
1) When we start the job the first time there will be thousands of Veterans to query. Similarly if the job ever fails for multiple days we'll have a backlog of Veterans to schedule. Spreading this out over hour chunks will ensure we don't hit any rate limits, nor overwhelm any dependencies.

Query for the Appeals we need to schedule.
```
vacols_ids = VACOLS::Case.where(bfcurloc: 57).pluck(:id)

# Find VACOLS cases that don't have appeals in our system.
existing_vacols_ids = LegacyAppeal.where(vacols_id: vacols_ids).pluck(:vacols_id)
missing_vacols_ids = vacols_ids - existing_vacols_ids

# Create missing legacy appeals
missing_vacols_ids.each do |vacols_id| # Schedule no more than 500/hour
  LegacyAppeal.find_or_create_by_vacols_id(vacols_id)
end

appeals = LegacyAppeal.where(vacols_id: vacols_ids).includes(:hearing_locations).where(hearing_location: { id: nil }).limit(500)

appeals.each do |appeal|
  veteran = appeal.veteran
  lat, long = ExternalApi::VADotGovService.geocode(
    address_line_1: veteran.address_line1,
    address_line2: veteran.address_line2,
    address_line3: veteran.address_line3,
    city: veteran.city,
    state: veteran.state,
    country: veteran.country,
    zip_code: veteran.zip_code
  )

  ids = ids_for_ro # to be defined
  ExternalApi::VADotGovService(lat, long, ids).each do |alternate_hearing_location|
    AlternateHearingLocation.create(appeal_id: appeal.id, **alternate_hearing_location)
  end
end
```

**Comment by mdbenjam** [comment]
Open question, might we ever need to redo the AHL query for a veteran? to update it?

Also I realize that `VACOLS::Case.where(bfcurloc: 57).limit(500)` is not sufficient. We need to also make sure these appeals do not have AHL yet. When do appeals get schedule hearing tasks? It would be easier to do this if we used tasks instead of locations...

**Comment by mdbenjam** [comment]
Updated to handle the fact that we may not already have the appeals in our DB and might have to create them first.

**Comment by lomaxap** [comment]
I think it's currently an open question whether AHL should be associated with a veteran or an appeal.

**Comment by mdbenjam** [comment]
So originally I was thinking we should associate it with a Veteran, and still think there's a compelling case for it. However, I thought appeal might be better to avoid questions of when we need to invalidate the cache. If we do it on a per appeal basis, then if a new appeal from the same Veteran comes to the Board we can look up the best hearing location for them at their current address, while not touching the records we'd already saved from the previous iteration. If we do it on a per Veteran level, it seems like we'll either need to clear existing records associated with the Veteran when they come back for a new appeal, or we will need to mark them as no longer relevant.

Would love to hear arguments for attaching it to the Veteran table.

**Comment by lomaxap** [comment]
For context, [here's the original tech spec Sharon and I drafted.](https://github.com/department-of-veterans-affairs/appeals-design-research/blob/master/Projects/Hearings/Hearing%20Schedule/Tech%20Specs/AlternateHearingLocations.md)

I originally thought there were scenarios where an RO could change, and we'd need to invalidate it (but @sharon thinks that probably will never happen). I was also considering times we may need to invalidate when a veteran changes address. The closest AHL would change, but maybe not the RO. That would probably happen infrequently, but is worth noting. 

The main motivation for me is that we would need to duplicate this logic for LegacyAppeal and Appeal models. Putting it on the Veteran simplifies the implementation.

**Comment by mdbenjam** [comment]
> The main motivation for me is that we would need to duplicate this logic for LegacyAppeal and Appeal models. Putting it on the Veteran simplifies the implementation.

I'm not sure I agree. We can make the `hearing_location` table have a polymorphic association with both legacy and ama appeals.

My concern around their address changing is not so much between the appeal being created and scheduling a hearing, but between two appeals. If someone has an appeal in 2019 and 2022 and has moved in-between, what do we do with the old save hearing locations connected to the veteran record?

**Comment by mdbenjam** [comment]
Another question is Redis or Postgres. I would suggest postgres since if we lose the cached info, we're hosed. It'll take an entire day to build the info back up into our cache, and that means 12 hours of people not scheduling hearings.

**Comment by lomaxap** [comment]
The hearing location and RO would still be saved on the actual Hearing. I'm currently thinking of the RO and AHL on the Veteran or Appeal as more of a default setting than a datapoint that we need a historical record of. We are only using it to work out which RO to associate a hearing with. I'm not sure it has any other use in AMA world (for legacy appeals, ROs on appeals were meaningful because they represented where the appeal's paperwork was processed. It seems less meaningful now). 

In terms of a record, schedulers will have the ability to change the hearing location and RO for a given hearing, so the data on the Appeal wouldn't necessarily reflect where all of the hearings took place.

> what do we do with the old save hearing locations connected to the veteran record?

In cases where an address changes or an RO changes, if we saved it on the Veteran, we would always overwrite the existing data. 

>  I think the code for the two cases will be fairly different because tasks and locations are so different :/. We can build a shared concern or service for any overlap.

I'm not following this point. If we saved it on the Veteran, wouldn't that mean we wouldn't have to write a shared concern?

> I would suggest postgres since if we lose the cached info, we're hosed.

I agree, and I think Sharon would too. I think we initially wanted to use Redis for expediency.
